File: python/day_1.py
from collections import Counter
from timeit import timeit

list_1 = []
list_2 = []

with open("../data/day_1", "r") as f:
    for line in f.readlines():
        split = line.split("   ")
        list_1.append(int(split[0].strip()))
        list_2.append(int(split[1].strip()))

list_1.sort()
list_2.sort()


def get_total_distance() -> int:
    distance = []
    for i in range(0, len(list_1)):
        distance.append(abs(list_1[i] - list_2[i]))

    return sum(distance)


def get_similarity() -> int:
    similarity = 0
    occurances = Counter(list_2)
    for value in list_1:
        similarity += value * occurances[value]

    return similarity


def benchmark():
    get_total_distance()
    get_similarity()

    assert get_total_distance() == 1530215
    assert get_similarity() == 26800609


result = timeit("benchmark()", setup="from __main__ import benchmark", number=10000)

avg = result / 10000

print(f"Time: {avg:.10f}")

File: rust/src/day_1.rs
use std::{fs, io};
pub struct LocationList {
    pub list_1: Vec<u32>,
    pub list_2: Vec<u32>,
}

impl LocationList {
    pub fn build() -> io::Result<Self> {
        let mut list_1: Vec<u32> = Vec::with_capacity(1000);
        let mut list_2: Vec<u32> = Vec::with_capacity(1000);

        fs::read_to_string("../data/day_1")?
            .lines()
            .map(|x| x.split_whitespace().collect::<Vec<_>>())
            .for_each(|s| {
                list_1.push(s[0].parse().unwrap());
                list_2.push(s[1].parse().unwrap());
            });

        list_1.sort();
        list_2.sort();

        Ok(Self { list_1, list_2 })
    }
    #[inline]
    pub fn get_total_distance(&self) -> u32 {
        self.list_1
            .iter()
            .enumerate()
            .map(|(i, dist)| dist.abs_diff(self.list_2[i]))
            .sum()
    }
    pub fn get_similarity(&self) -> u32 {
        let mut list_2_occurances = vec![0u32; *self.list_1.iter().max().unwrap() as usize + 1];

        self.list_2
            .iter()
            .fold(&mut list_2_occurances, |acc, &idx| {
                (*acc)[idx as usize] = (*acc)[idx as usize].saturating_add(1);
                acc
            });

        self.list_1
            .iter()
            .map(|x| *x * list_2_occurances[*x as usize])
            .sum()
    }
}
